### PR TITLE
Automatically export new episodes to Kodi library

### LINKED ISSUE #316: Auto-add new episodes of exported shows to library
*I'm submitting a ...*
  - [x] feature request

This is more of a question: Would you be willing to merge a feature, that periodically checks for new episodes of shows that were exported to the library and automatically exports/adds them?

I've currently got this in the works and will be using it personally. If this feature were welcome, I'll clean up the code and make PR.

### PR DESCRIPTION
**Pull request by caphm**
This adds a schedule for checking for new episodes of shows exported to Kodi library and automatically exporting them (Closes #316). When new episodes were added to the library, a notification is shown to the user. Apart from this, the operation runs completely in the background.

**Important workaround:**
To check for new episodes, the netflix id of the show is required. It is now being added to the addon-internal library when exporting a show. Shows already exported before are missing this however, so the feature won't work for these out of the box. If this case is encountered, an error is written to the log. To fix it, just remove and re-add the show to your Kodi library (these  instructions are also printed to the log).

Behaviour is configurable via the addon settings under the `Library` category:
- Schedule interval (never, daily, every other day, every 5 days, weekly) - disable/re-enable of addon is required when changing this
- time of day when new episodes should be added
- only update library when idle: If enabled, will defer updates until Kodi has been idle for >= 5 minutes
- Checking interval to determine if schedule conditions have been met

The check interval is for performance reasons, because we do not need to check every couple seconds when going through the service main loop. This defers the actual schedule check to longer intervals.

**Conflicts with other PRs:**
I had to rewrite a lot of service.py, so this will conflict with #293. When the other PR gets merged, I'll do a rebase, just let me know.

I have also added some new strings to the language files, but know english and german. Anybody wanting to do an update for the others is welcome to do so :)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the [**CONTRIBUTING**](Contributing.md) document.

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by hawkeyexp** [comment]
I have done a test merge and it seems it breaks export of strm files - no error but no export to destination folder

**Comment by caphm** [comment]
Does it break exporting in general (i.e. by using the already implemented way via 'Export to Library') or only when doing it automatically?
I have encountered this once while testing, but was never able to reproduce it, so I thought it was just a hiccup.

Can you describe the steps you took when this occured? Which show were you exporting, was it already in your library before, etc...

**Comment by hawkeyexp** [comment]
It has done a general break - tested with existing show like stranger things which i have updated (my routine removes all and does a re export while updating via contenxt menu).
also tried with new show which was not exported before.
The way is the default one - right click on show to export to library

**Comment by caphm** [comment]
That's strange, I cannot repduce the issue at all. Tested with a clean install, library and no previous addon data. I tried Stranger Things and several other shows, everything works fine.

Did you also merge other PRs into the same branch? This one's a direct descendant of up-to-date master, maybe there are some conflicts with other PRs.

Also, are only the strm-files missing but folders get created? What are the contents of lib.ndb in netflix addon_data directory?

I will push a change to add more debug logging, maybe we can pinpoint the issues this way.

**Comment by hawkeyexp** [comment]
will check it again later today

**Comment by caphm** [comment]
I've added a lot of debug logging calls. A successful run should output something like this: https://pastebin.com/mzJNt6PX

**Comment by hawkeyexp** [comment]
I tried again and it seems to work correct - possible a bug on first test merges :-(
I also noticed a conflict with my extendview pr and commited a change to fix the conflict.
Ony have to check if new episodes will be auto export.

**Comment by caphm** [comment]
Good to know it works! We might need to keep an eye on wether this issue pops up again. 

To simplify testing, you can change the return line of library_update_scheduled (line 188 in service.py) to always return true. Otherwise you might need to wait an entire day for the update to be triggered ;)

**Comment by joaosagrath** [comment]
Let me kidnap this conversation to suggest a feature that is directly linked to Auto Update Episodes.

How about an option where all the items present in "My List" would be added to the library? If we added an item to "my list" the addon would export that item automatically. And if we remove any item, the addon would automatically delete.

This option could be of an on / off type and could follow the same rules that follow in this feature of updating the episodes.

We currently have to export "My List" items one by one. For those who have several items, it will be necessary several actions to export.

I do not know if they agree with me, but if someone adds something to "My List" it is because they intend to attend at some point, right?

**Comment by hawkeyexp** [comment]
The base idea is good but personally i won't use it cause for example i use my list as marker and transfer platform between devices. Not all things in my list are interesting (good enough) to add to my kodi db.

**Comment by caphm** [comment]
Already working on it ;) Had the same idea when I had to remove and re-add all shows in my library to work with this feature. I have some ideas in which direction this could go (one-way, two-way sync, etc), but this would clutter up the conversation of this PR. Would you open a feature request so we can move the conversation there?

Regarding removing and re-adding shows: I might have found a way to automatically "migrate" the internal DB (i.e. adding netflix IDs to shows) that does not require the user to remove and re-add every show. Still testing this though...

**Comment by raven6679** [comment]
Let me jump in and stir a little ;) even if it's not related 100% to the original topic. I think groundbreaker here would be 2way sync between Kodi <> Netflix, and I'm talking about stuff like reccomendation on Netflix account based on episodes/movies watched in Kodi using Netflix plugin...

File: resources/lib/Navigation.py [hunk 1]
```diff
@@ -608,6 +612,25 @@ def remove_from_library(self, video_id, season=None, episode=None):
         self.kodi_helper.dialogs.show_no_metadata_notify()
         return False
 
+    @log
+    def export_new_episodes(self):
+        no_errors = True
+        for title, meta in self.library.list_exported_shows().iteritems():
+            try:
+                self.log('Exporting new episodes of {} (id={})'
+                         .format(title, meta['netflix_id']))
```
Comment (hawkeyexp) [inline resources/lib/Navigation.py]: title and alt_title should be encoded as utf-8 cause it results in errors on special chars like in "marvel's iron fist" in my case - i also have some other things found and figure out what goes wrong

Comment (caphm) [reply resources/lib/Navigation.py]: Yeah, I've noticed that on other accounts. Unicode handling in 2.7 is a pain :(
Will you make changes to the PR or should I push an update?

**Comment by caphm** [comment]
@raven6679, I'm with you in wanting to have such a feature set, but unfortunately this won't be possible until someone figures out how to report watched status back to netflix. There's also an open feature request for this, so please move the conversation there and keep it focused on the PR here ;)

**Comment by joaosagrath** [comment]
@caphm 
open the feature request
is #317

File: resources/lib/Navigation.py (hunk 1 above)
Comment (hawkeyexp) [reply resources/lib/Navigation.py]: Push an update please

Comment (hawkeyexp) [reply resources/lib/Navigation.py]: Ignore it - i will push a pr to you if all is fixed - found another unicode bug :-( It sucks...

**Comment by caphm** [comment]
Thanks for fixing the unicode issues.

I've added code that fetches the netflix ID of a show if it's missing in the internal DB. This way you don't have to remove and re-add shows to work with this feature.

I think we should be good to go for a merge by now :)

**Pull request merged**